pkg/auth/auth.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package auth
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"time"
+
+	"github.com/casbin/casbin/v2"
+	"github.com/casbin/casbin/v2/model"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type authRecord struct {
+	role   string
+	expiry time.Time
+}
+
+type apiKeys map[string]authRecord
+
+type Authenticator struct {
+	username     string
+	passwordHash []byte
+	keys         apiKeys
+	expires      time.Duration
+	enforcer     *casbin.Enforcer
+}
+
+func New(username, password string, expires time.Duration) (*Authenticator, error) {
+	m, err := model.NewModelFromString(`
+	[request_definition]
+	r = sub, obj, act
+	
+	[policy_definition]
+	p = sub, obj, act
+	
+	[policy_effect]
+	e = some(where (p.eft == allow))
+	
+	[matchers]
+	m = r.sub == p.sub && keyMatch(r.obj, p.obj) && regexMatch(r.act, p.act)`)
+
+	if err != nil {
+		return nil, err
+	}
+
+	e, err := casbin.NewEnforcer(m)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := applyPolicies(e); err != nil {
+		return nil, err
+	}
+
+	passwordHash, err := bcrypt.GenerateFromPassword([]byte(password), 14)
```

Comment:
I had in mind to avoid keeping in configuration password in plain text by using bcrypt, but this is not doing much for that as it is still required to set password in plain text in config in order to hash it here. The hash should be set in the configuration, instead the plain password. The hash may be generated by a separate command or a tool like htpasswd or any online bcrypt site. I think that the cli command would be nice. It does add a ux complexity, but I think for a justified security reason.

Reply:
should we maybe add it as a separate command?
`/cmd/bee` <- current entry point
`/cmd/passwordgen` <- new command that takes a plain text and returns its hash

or maybe just document so the user could use existing tools?

Reply:
My opinion is that a separate command would provide a direct way to obtain the hash, and to have in documentation how a password can be generated. But for this PR, the command is not necessary, only documenting where the hash can be  generated. For example, nginx relied on apache's htpasswd tool for many years and clearly documented how to use it in relation to nginx. So I would not rush with the command, but I think that it provides a nice convenience.